File: snail.go
package kata

// Snail ...
func Snail(snaipMap [][]int) []int {
	l := len(snaipMap[0])
	if l == 0 {
		return []int{}
	} else if l == 1 {
		return []int{snaipMap[0][0]}
	}

	h, v := 0, 0
	var result []int

	for h = 0; h < l; h++ {
		result = append(result, snaipMap[0][h])
	}
	for v = 1; v < l; v++ {
		result = append(result, snaipMap[v][l-1])
	}
	for h = l - 2; h >= 0; h-- {
		result = append(result, snaipMap[l-1][h])
	}
	for v = l - 2; v > 0; v-- {
		result = append(result, snaipMap[v][0])
	}

	if l > 2 {
		inner := [][]int{}

		for _, e := range snaipMap[1 : l-1] {
			inner = append(inner, e[1:l-1])
		}

		result = append(result, Snail(inner)...)
	}

	return result
}

File: snail.js
let snail = (array) => {
	let result;
	while (array.length) {
		result = (result ? result.concat(array.shift()) : array.shift());
		for (let i = 0; i < array.length; i++) {
			result.push(array[i].pop());
		}
		result = result.concat((array.pop() || []).reverse());
		for (let i = array.length - 1; i >= 0; i--) {
			result.push(array[i].shift());
		}
	}
	return result;
};

// Refactored

snail = (array) => {
	var result;
	while (array.length) {
		// Steal the first row.
		result = (result ? result.concat(array.shift()) : array.shift());
		// Steal the right items.
		for (var i = 0; i < array.length; i++)
			result.push(array[i].pop());
		// Steal the bottom row.
		result = result.concat((array.pop() || []).reverse());
		// Steal the left items.
		for (let i = array.length - 1; i >= 0; i--)
			result.push(array[i].shift());
	}
	return result;
};